waiter/src/waiter/auth/saml.clj

```diff
@@ -0,0 +1,195 @@
+;;
+;; Copyright (c) Two Sigma Open Source, LLC
+;;
+;; Licensed under the Apache License, Version 2.0 (the "License");
+;; you may not use this file except in compliance with the License.
+;; You may obtain a copy of the License at
+;;
+;;  http://www.apache.org/licenses/LICENSE-2.0
+;;
+;; Unless required by applicable law or agreed to in writing, software
+;; distributed under the License is distributed on an "AS IS" BASIS,
+;; WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+;; See the License for the specific language governing permissions and
+;; limitations under the License.
+;;
+(ns waiter.auth.saml
+  (:require [clj-time.core :as t]
+            [clojure.data.codec.base64 :as b64]
+            [clojure.java.io :as io]
+            [clojure.tools.logging :as log]
+            [comb.template :as template]
+            [ring.middleware.params :as ring-params]
+            [saml20-clj.routes :as saml-routes]
+            [saml20-clj.shared :as saml-shared]
+            [saml20-clj.sp :as saml-sp]
+            [taoensso.nippy :as nippy]
+            [waiter.auth.authentication :as auth]
+            [waiter.middleware :as middleware]
+            [waiter.util.utils :as utils]))
+
+(defrecord SamlAuthenticator [auth-redirect-uri idp-cert idp-uri password saml-acs-handler-fn saml-auth-redirect-handler-fn saml-req-factory!]
+  auth/Authenticator
+  (wrap-auth-handler [_ request-handler]
+    (fn saml-authenticator-handler [{:keys [headers query-string request-method uri] :as request}]
+      (let [waiter-cookie (auth/get-auth-cookie-value (get headers "cookie"))
+            [auth-principal _ :as decoded-auth-cookie] (auth/decode-auth-cookie waiter-cookie password)]
+        (cond
+          ;; Use the cookie, if not expired
+          (auth/decoded-auth-valid? decoded-auth-cookie)
+          (let [auth-params-map (auth/auth-params-map auth-principal)
+                request-handler' (middleware/wrap-merge request-handler auth-params-map)]
+            (request-handler' request))
+          :else
+          (case request-method
+            :get (let [saml-request (saml-req-factory!)
+                       relay-state (str (name (utils/request->scheme request)) "://" (get headers "host") uri (if query-string "?" "") query-string)]
+                   (saml-sp/get-idp-redirect idp-uri saml-request relay-state))
+            :post (if-let [saml-auth-data (get-in (ring-params/params-request request) [:form-params "saml-auth-data"])]
+                    (let [{:keys [not-on-or-after saml-principal]}
+                          (try
+                            (nippy/thaw (b64/decode (.getBytes saml-auth-data))
+                                        {:password password :v1-compatibility? false :compressor nil})
+                            (catch Exception e
+                              (throw (ex-info "Could not parse saml-auth-data." {:status 400
```

Comment:
pass `e` as the third argument to `ex-info` so we can preserve the cause.

Reply:
done